File: solutions/141.linked-list-cycle/tow-pointers.ts
/*
 * @lc app=leetcode.cn id=141 lang=typescript
 *
 * https://leetcode-cn.com/problems/linked-list-cycle
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

// @lc code=start
function hasCycle(head: ListNode | null): boolean {
    if (!head) {
        return false;
    }
    let slow = head;
    let fast = head;
    while (fast.next && fast.next.next) {
        slow = slow.next!;
        fast = fast.next.next;
        if (slow === fast) {
            return true;
        }
    }
    return false;
}
// @lc code=end

export {};

File: solution/26.remove-duplicates-from-sorted-array/tow-pointers.rs
/*
 * @lc app=leetcode.cn id=26 lang=rust
 *
 * https://leetcode-cn.com/problems/remove-duplicates-from-sorted-array
 */

// @lc code=start
impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        if (nums.len() == 0) {
            return 0;
        }
        let mut i = 0;
        let mut j = 1;
        while j < nums.len() {
            if nums[i] == nums[j] {
                j += 1;
            } else {
                i += 1;
                nums[i] = nums[j];
                j += 1;
            }
        }
        (i + 1) as i32
    }
}
// @lc code=end
